src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -261,166 +290,227 @@ private SendingReport sendEvents(CrawlerId crawlerId, DataProcessorInfo dataProc
+
+                if (!recoveryStateMessages.isEmpty()) {
+                    RecoveryState newState;
+
+                    if (oldState == null) {
+                        Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> lastProcessedMessages = new HashMap<>(recoveryStateMessages);
+
+                        newState = new RecoveryState(null, lastProcessedMessages,
+                                0,
+                                numberOfMessages);
+                    } else {
+                        Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> lastProcessedMessages = new HashMap<>();
+
+                        if (oldState.getLastProcessedMessages() != null) {
+                            lastProcessedMessages.putAll(oldState.getLastProcessedMessages());
+                        }
+
+                        lastProcessedMessages.putAll(recoveryStateMessages);
+
+                        newState = new RecoveryState(oldState.getLastProcessedEvent(), lastProcessedMessages,
+                                oldState.getLastNumberOfEvents(),
+                                numberOfMessages);
+                    }
+
+                    interval = intervalsWorker.updateRecoveryState(interval, newState.convertToJson());
                 }
             }
 
             search = messages.size() == batchSize;
         }
 
-        return new SendingReport(CrawlerAction.NONE, dataServiceName, dataServiceVersion);
+        return new SendingReport(CrawlerAction.NONE, interval, dataProcessorName, dataProcessorVersion, 0, numberOfMessages);
+    }
+
+    private Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> findPairedMessages
+            (Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> messages, Timestamp from, Timestamp to,
+             MessageID resumeId, Collection<String> aliases) {
+        Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> result = new HashMap<>();
+
+        messages.forEach((key, value) -> {
+
+            com.exactpro.th2.common.grpc.Direction direction = key.getDirection();
+            String alias = key.getSessionAlias();
+
+            RecoveryState.InnerMessage message = CrawlerUtils.findPairedMessage(value, messages);
+
+            if (message != null) {
+                result.put(new CrawlerUtils.AliasAndDirection(alias, direction), message);
+            } else {
+                MessageSearchRequest.Builder searchBuilder = MessageSearchRequest.newBuilder();
+
+                int limit = 1;
+
+                Iterator<StreamResponse> iterator = CrawlerUtils.searchMessages(dataProviderService,
+                        new CrawlerUtils.MessagesSearchInfo(searchBuilder, from, to, limit, resumeId, aliases));
+
+                while (iterator.hasNext()) {
+                    iterator.next().getStreamInfo().getStreamsList().forEach(stream -> {
```

Comment:
Why do you think the first result in the iterator will be the `StreamInfo`? It might be a message because you use the limit 1